File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        for _ in range(n):
           fast = fast.next
        
        slow = head

        #Edge case when N = 5. In this case fast will bcome None.
        #To understand better, dry run on 1-2-3-4-5 for N=5
        if not fast:
            return head.next
        
        while fast.next:
            slow = slow.next
            fast = fast.next 
        
        slow.next = slow.next.next
        return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode curr = head;
        int len = getLength(head);

        if(len-n == 0){
            return head.next;
        }

        for(int i = 1; i < len-n; i++){
            curr = curr.next;
        }

        curr.next = curr.next.next;
        return head;
    }

    private int getLength(ListNode head){
        ListNode curr = head;
        int len = 0;
        while(curr != null){
            len++;
            curr = curr.next;
        }
        return len;
    }
}